### PR TITLE
Add symbol table

### PR DESCRIPTION
**Pull request by ericmercer**
The symbol table builds from a text description of the state of the workflow. It includes types and restrictions on values. The syntax is defined in an ANTLR grammar. The symbol table itself is built after parsing the text desctription of the state with a listner using the ANTLR runtime. Building a symbol table includes semantic checks:

  -   No multiple definitions
  -   If values are restricted for a variable, then the initial value must be included in the restricted set
  -   Initial values for const types must be assignment compatible with the declared type
  -   Initial values and restricted values for variables must be assignment compatible with the declared type

The commit includes the base machinery for type checking.

### CODE REVIEW

File: src/bpmncwpverify/antlr/State.interp [hunk 1]
```diff
@@ -24,10 +34,13 @@ WS
...
... 0, 17, 15, 1, 0, 0, 0, 17, 18, 1, 0, 0, 0, 18, 23, 1, 0, 0, 0, 19, 17, 1, 0, 0, 0, 20, 22, 3, 6, 3, 0, 21, 20, 1, 0, 0, 0, 22, 25, 1, 0, 0, 0, 23, 21, 1, 0, 0, 0, 23, 24, 1, 0, 0, 0, 24, 27, 1, 0, 0, 0, 25, 23, 1, 0, 0, 0, 26, 28, 3, 8, 4, 0, 27, 26, 1, 0, 0, 0, 28, 29, 1, 0, 0, 0, 29, 27, 1, 0, 0, 0, 29, 30, 1, 0, 0, 0, 30, 31, 1, 0, 0, 0, 31, 32, 5, 0, 0, 1, 32, 1, 1, 0, 0, 0, 33, 34, 5, 6, 0, 0, 34, 35, 5, 13, 0, 0, 35, 36, 5, 10, 0, 0, 36, 37, 3, 4, 2, 0, 37, 38, 5, 11, 0, 0, 38, 3, 1, 0, 0, 0, 39, 41, 5, 13, 0, 0, 40, 39, 1, 0, 0, 0, 41, 42, 1, 0, 0, 0, 42, 40, 1, 0, 0, 0, 42, 43, 1, 0, 0, 0, 43, 5, 1, 0, 0, 0, 44, 45, 5, 5, 0, 0, 45, 46, 5, 13, 0, 0, 46, 47, 5, 1, 0, 0, 47, 48, 3, 10, 5, 0, 48, 49, 5, 9, 0, 0, 49, 50, 5, 13, 0, 0, 50, 7, 1, 0, 0, 0, 51, 52, 5, 12, 0, 0, 52, 53, 5, 13, 0, 0, 53, 54, 5, 1, 0, 0, 54, 55, 3, 10, 5, 0, 55, 56, 5, 9, 0, 0, 56, 61, 5, 13, 0, 0, 57, 58, 5, 10, 0, 0, 58, 59, 3, 4, 2, 0, 59, 60, 5, 11, 0, 0, 60, 62, 1, 0, 0, 0, 61, 57, 1, 0, 0, 0, 61, 62, 1, 0, 0, 0, 62, 9, 1, 0, 0, 0, 63, 66, 3, 12, 6, 0, 64, 66, 5, 13, 0, 0, 65, 63, 1, 0, 0, 0, 65, 64, 1, 0, 0, 0, 66, 11, 1, 0, 0, 0, 67, 68, 7, 0, 0, 0, 68, 13, 1, 0, 0, 0, 6, 17, 23, 29, 42, 61, 65]
```
Comment (jminyoon) [inline src/bpmncwpverify/antlr/State.interp:46]: No changes needed just confused what this is for. If you see this, just send a message through Discord and explain to me what this is for if you can

File: src/bpmncwpverify/state.py [hunk 2]
```diff
@@ -1,37 +1,30 @@
-from antlr4 import CommonTokenStream, InputStream
+from antlr4 import CommonTokenStream, InputStream, ParseTreeWalker
```
Comment (jminyoon) [inline src/bpmncwpverify/state.py:1]: All of the symbol table code looks clean to me. I am curious if you think this file is too large and if we should split the functionality into different files

File: src/bpmncwpverify/typechecking.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+from returns.result import Failure, Result, Success
```
Comment (jminyoon) [inline src/bpmncwpverify/typechecking.py:1]: Should I be using this once I pull it down for my expression type checker?

**Review by jminyoon — COMMENTED** [review]
Overall looks good to me. If you could just answer the questions I left in the comments

File: src/bpmncwpverify/antlr/State.interp (hunk 1 above)
Comment (ericmercer) [reply src/bpmncwpverify/antlr/State.interp:46]: That is part of the ANTLR generated code. I really don't know what it is for either though I suspect it encodes some type of transition table for the parser.

File: src/bpmncwpverify/state.py (hunk 2 above)
Comment (ericmercer) [reply src/bpmncwpverify/state.py:1]: Are you referring to the `ParseTreeWalker` or `state.py`. If it's the `ParseTreeWalker`, then that is an ANTLR generated file that we effectively use and don't look at. The `state.py` file is only 206 LOC.

File: src/bpmncwpverify/typechecking.py (hunk 3 above)
Comment (ericmercer) [reply src/bpmncwpverify/typechecking.py:1]: Yes. It computes the types for literals, and it provides the type-check for assignment compatibility. It needs to be expanded with other compatibility checks: logical connectives and relational operators.

**Pull request merged**